grpc/src/main/java/com/linecorp/armeria/server/grpc/ProtobufRequestConverterFunction.java

```diff
@@ -0,0 +1,195 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.grpc;
+
+import static java.lang.invoke.MethodType.methodType;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.ParameterizedType;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.MapMaker;
+import com.google.protobuf.ExtensionRegistry;
+import com.google.protobuf.Message;
+import com.google.protobuf.util.JsonFormat;
+import com.google.protobuf.util.JsonFormat.Parser;
+
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.annotation.RequestConverterFunction;
+
+/**
+ * A {@link RequestConverterFunction} which converts a Protocol Buffers or JSON body of
+ * the {@link AggregatedHttpRequest} to an object.
+ * The built-in parser of {@link Message} for Protocol Buffers is applied only when the {@code content-type} of
+ * {@link RequestHeaders} either {@link MediaType#PROTOBUF} or {@link MediaType#OCTET_STREAM} or
+ * the {@link MediaType#subtype()} contains {@code "protobuf"}.
+ * The {@link Parser} for JSON is applied only when the {@code content-type} of
+ * the {@link RequestHeaders} is {@link MediaType#JSON} or ends with {@code +json}.
+ *
+ * <p>The Protocol Buffers spec does not have an official way to sending multiple messages because
+ * an encoded message does not have self-delimiting.
+ * See
+ * <a href="https://developers.google.com/protocol-buffers/docs/techniques#streaming">Streaming Multiple Messages</a>
+ * for more information.
+ * Therefore a sequence of Protocol Buffer messages can not be handled by this {@link RequestConverterFunction}.
+ * However a <a href="https://tools.ietf.org/html/rfc7159#section-5">JSON array</a> body
+ * could be converted into a {@link Collection} type. e.g, {@code List<Message>} and {@code Set<Message>}.
+ */
+public final class ProtobufRequestConverterFunction implements RequestConverterFunction {
```

Comment:
Shouldn't we make a separate module so that the user doesn't have to bring gRPC dependencies to use this?

Reply:
I thought this feature would be interesting to some users who want to gRPC service and client already. Do we need `armeria-protobuf`? 🤔

Reply:
I'm not sure. 🤣  @anuraaga Do you have any suggestions?

Reply:
Sorry for late reply - it does seem strange to have this in gRPC package since it seems to have nothing to do with gRPC. Perhaps core with compileOnly dependency on protobuf so it's next to the other annotated service functionality? not sure

Reply:
Yeah, the code is too small to create a seperate module. Let me move this code to core and add `@UnstableApi` in case we need `armeria-protobuf` later.